File: ClipCascade_Mobile/src/SetupBundle.js
const SETUP_BUNDLE_TYPE = 'clipcascade-setup-v1';
const SYNC_KEY_PATTERN = /^ccsk_[A-Za-z0-9_-]{43}$/;

export const parseSetupBundle = rawBundle => {
  if (!rawBundle || String(rawBundle).trim() === '') {
    throw new Error('Setup bundle is empty');
  }

  let bundle;
  try {
    bundle = JSON.parse(String(rawBundle));
  } catch (error) {
    throw new Error('Setup bundle is not valid JSON');
  }

  if (!bundle || typeof bundle !== 'object' || Array.isArray(bundle)) {
    throw new Error('Setup bundle must be a JSON object');
  }
  if (bundle.type !== SETUP_BUNDLE_TYPE) {
    throw new Error('Unsupported setup bundle type');
  }

  [
    'serverUrl',
    'username',
    'apiKey',
    'clientId',
    'clientName',
  ].forEach(key => {
    if (!String(bundle[key] || '').trim()) {
      throw new Error(`Setup bundle is missing ${key}`);
    }
  });

  if (bundle.encryptionMode === 'sync_key') {
    const syncKey = String(bundle.syncEncryptionKey || '').trim();
    if (!SYNC_KEY_PATTERN.test(syncKey)) {
      throw new Error('Setup bundle has an invalid sync encryption key');
    }
  }

  return bundle;
};

export const syncEncryptionKeyToBase64 = syncEncryptionKey => {
  const syncKey = String(syncEncryptionKey || '').trim();
  if (!SYNC_KEY_PATTERN.test(syncKey)) {
    throw new Error('Sync encryption key must be a valid ccsk_ key');
  }

  const base64Url = syncKey.slice(5);
  return `${base64Url.replace(/-/g, '+').replace(/_/g, '/')}=`;
};

export const applySetupBundleToData = (data, rawBundle) => {
  const bundle = parseSetupBundle(rawBundle);
  return {
    ...data,
    server_url: String(bundle.serverUrl).trim().replace(/\/+$/, ''),
    username: String(bundle.username).trim(),
    api_key: String(bundle.apiKey).trim(),
    api_client_id: String(bundle.clientId).trim(),
    api_client_name: String(bundle.clientName).trim(),
    cipher_enabled: String(bundle.cipherEnabled !== false),
    sync_encryption_key:
      bundle.encryptionMode === 'sync_key'
        ? String(bundle.syncEncryptionKey).trim()
        : '',
    hashed_password: '',
    csrf_token: '',
  };
};

File: ClipCascade_Desktop/src/utils/setup_bundle.py
import json
import re


SETUP_BUNDLE_TYPE = "clipcascade-setup-v1"
SYNC_KEY_PATTERN = re.compile(r"^ccsk_[A-Za-z0-9_-]{43}$")


def parse_setup_bundle(raw_bundle: str) -> dict:
    if not raw_bundle or not raw_bundle.strip():
        raise ValueError("Setup bundle is empty")

    try:
        bundle = json.loads(raw_bundle)
    except json.JSONDecodeError as exc:
        raise ValueError("Setup bundle is not valid JSON") from exc

    if not isinstance(bundle, dict):
        raise ValueError("Setup bundle must be a JSON object")
    if bundle.get("type") != SETUP_BUNDLE_TYPE:
        raise ValueError("Unsupported setup bundle type")

    for key in ("serverUrl", "username", "apiKey", "clientId", "clientName"):
        if not str(bundle.get(key) or "").strip():
            raise ValueError(f"Setup bundle is missing {key}")

    if bundle.get("encryptionMode") == "sync_key":
        sync_key = str(bundle.get("syncEncryptionKey") or "").strip()
        if not SYNC_KEY_PATTERN.match(sync_key):
            raise ValueError("Setup bundle has an invalid sync encryption key")

    return bundle


def apply_setup_bundle_to_config(config, raw_bundle: str) -> dict:
    bundle = parse_setup_bundle(raw_bundle)

    config.data["server_url"] = str(bundle["serverUrl"]).strip().rstrip("/")
    config.data["username"] = str(bundle["username"]).strip()
    config.data["api_key"] = str(bundle["apiKey"]).strip()
    config.data["api_client_id"] = str(bundle["clientId"]).strip()
    config.data["api_client_name"] = str(bundle["clientName"]).strip()
    config.data["cipher_enabled"] = bool(bundle.get("cipherEnabled", True))
    config.data["password"] = ""
    config.data["hashed_password"] = None
    config.data["cookie"] = None
    config.data["csrf_token"] = ""

    if bundle.get("encryptionMode") == "sync_key":
        config.data["sync_encryption_key"] = str(bundle["syncEncryptionKey"]).strip()

    return bundle
